File: src/Domain.Linnapps/CartonType.cs
namespace Linn.Products.Domain.Linnapps
{
    using Linn.Products.Domain.Linnapps.Exceptions;

    public class CartonType
    {
        public CartonType(string name, decimal width, decimal height, decimal depth)
        {
            if (string.IsNullOrWhiteSpace(name))
            {
                throw new IncompleteDataException("A carton type must have a name");
            }

            if (width <= 0 || height <= 0 || depth <= 0)
            {
                throw new IncompleteDataException("A carton type must have valid dimensions");
            }

            this.Name = name;
            this.Width = width;
            this.Height = height;
            this.Depth = depth;
        }

        private CartonType()
        {
            // ef
        }

        public string Name { get; set; }

        public string Description { get; set; }

        public decimal Height { get; set; }

        public decimal Width { get; set; }

        public decimal Depth { get; set; }

        public int NumberOfSmallLabels { get; set; }

        public int NumberOfLargeLabels { get; set; }
    }
}

File: src/Service.Host/client/src/components/__tests__/CartonType.js
import React from 'react';
import { createShallow } from '@material-ui/core/test-utils';
import CartonType from '../CartonType';

describe('<CartonType />', () => {
    let wrapper;
    let props;

    const getLoading = () => wrapper.find('Loading');
    const getInputFields = () => wrapper.find('InputField');
    const getButtons = () => wrapper.find('SaveBackCancelButtons');

    const shallow = createShallow({ dive: false });

    beforeEach(() => {
        props = {
            loading: true,
            editStatus: 'view',
            history: {}
        };

        wrapper = shallow(<CartonType {...props} />);
    });

    describe('when loading', () => {
        it('should render loading spinner', () => {
            expect(getLoading()).toHaveLength(1);
        });
    });

    describe('when carton type has loaded', () => {
        beforeEach(() => {
            wrapper.setProps({
                loading: false,
                cartonType: {
                    name: 'Carton 1',
                    numberOfSmallLabels: 1,
                    description: 'Carton description',
                    width: 3,
                    height: 2,
                    depth: 1,
                    links: [
                        {
                            rel: 'self',
                            href: '/products/maint/carton-types/1'
                        }
                    ]
                }
            });
        });

        it('should render text fields', () => {
            expect(getInputFields()).toHaveLength(5);
        });

        it('should render buttons', () => {
            expect(getButtons()).toHaveLength(1);
        });
    });
});
